### PR TITLE
cli: Rework the returned status codes

### PR DESCRIPTION
**Pull request by sschuberth**
The general rules are:

- Throw a UsageError (with a default status code of 1) to indicate that
  the user did something wrong.

- Throw a ProgramResult for errors or failures that are out of the
  user's control.

- Use a status code of 1 with ProgramResult for errors that prevented the
  program to execute correctly.

- Use a status code >= 2 with ProgramResult for failures found during
  successful execution, like rule violations.

Signed-off-by: Sebastian Schuberth <sebastian.schuberth@bosch.io>

### CODE REVIEW

**Comment by sschuberth** [comment]
@fviernau please double-check that this does not cause any issues in our Jenkins build steps.

File: cli/src/main/kotlin/commands/ScannerCommand.kt [hunk 1]
```diff
@@ -167,5 +168,17 @@ class ScannerCommand : CliktCommand(name = "scan", help = "Run existing copyrigh
             println("Writing scan result to '$file'.")
             file.mapper().writerWithDefaultPrettyPrinter().writeValue(file, ortResult)
         }
+
+        val scanResults = ortResult.scanner?.results
+
+        if (scanResults == null) {
```
Comment (fviernau) [inline cli/src/main/kotlin/commands/ScannerCommand.kt:174]: I tried to find a code path which, searching through `scanner.scanOrtResult()` and `scanner.scanPath`, which turns this ìf-condition` to true - I couldn't find any. Could you give me a hint?

Comment (sschuberth) [reply cli/src/main/kotlin/commands/ScannerCommand.kt:174]: I'm also pretty sure there is such code path that leaves `scanResults` with `null` after a scanner run. But the variable is nullable, so I need to handle the `null` case here somehow. And *if* it is `null` at this point, then certainly something did go very wrong.

Comment (fviernau) [reply cli/src/main/kotlin/commands/ScannerCommand.kt:174]: After looking into it again I believe it's fine to keep it this way.

One question popped up as I found that above mentioned functions do throw exception in case of some issues.
Do these exceptions have a similar effect as `throw ProgramResult(1)` and / or should they be caught by the comand?

Comment (sschuberth) [reply cli/src/main/kotlin/commands/ScannerCommand.kt:174]: I guess they should be caught and rethrown as exceptions recognized by *clikt* to set a status code accordingly. Which exceptions did you find to be thrown?

**Comment by fviernau** [comment]
I'm good with the PR `as-is`, not approving it yet as we need to align our internal setup first and test it a bit.

File: cli/src/main/kotlin/commands/AnalyzerCommand.kt [hunk 2]
```diff
@@ -148,5 +149,17 @@ class AnalyzerCommand : CliktCommand(name = "analyze", help = "Determine depende
...
         }
+
+        val analyzerResult = ortResult.analyzer?.result
+
+        if (analyzerResult == null) {
+            println("There was an error creating the analyzer result.")
+            throw ProgramResult(1)
+        }
+
+        if (analyzerResult.hasIssues) {
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/AnalyzerCommand.kt:160]: It just occurred to me that it would actually be nicer to only return `2` here if there are *unresolved* issues. However, the analyzer does not know currently whether an issue is unresolved or not, as we're not passing a `ResolutionProvider`. While that's certainly technically possible to do, this is a bigger change that I want to discuss in a bigger round first.

**Comment by fviernau** [comment]
@sschuberth I've aligned our setup and verified that it works after this PR is merged.
...so merging it now (as we discussed recently).

File: cli/src/main/kotlin/commands/ClearlyDefinedUploadCommand.kt [hunk 3]
```diff
@@ -153,7 +153,6 @@ class ClearlyDefinedUploadCommand : CliktCommand(
                 null -> println("failed to be uploaded.")
                 else -> {
                     println("was uploaded successfully:\n${summary.url}")
-
```
Comment (sschuberth) [inline cli/src/main/kotlin/commands/ClearlyDefinedUploadCommand.kt:156]: This change seems unrelated.

Comment (fviernau) [reply cli/src/main/kotlin/commands/ClearlyDefinedUploadCommand.kt]: Done.
